File: c++ version/1-100/39-combination-sum.cpp
vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<int> path;
    vector<vector<int>> ans;
    int sum = 0, startPos = 0;
    sort(candidates.begin(), candidates.end());
    backtracking(startPos, sum, candidates, target, path, ans);
    return ans;
}

void backtracking(int startPos, int sum, vector<int> &candidates, int target, vector<int> &path, vector<vector<int>> &ans) {
    if (sum == target) {
        ans.push_back(path);
        return;
    }

    for (int i = startPos; i < candidates.size(); ++i) {
        if (sum + candidates[i] > target) break;
        path.push_back(candidates[i]);
        sum += candidates[i];
        backtracking(i, sum, candidates, target, path, ans);
        sum -= candidates[i];
        path.pop_back();
    }
}

File: python version/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        curr = []
        ans = []
        n = len(candidates)

        def backtracking(start):
            if sum(curr) > target or start == n:
                return
            if sum(curr) == target:
                ans.append(curr[:])
                return
            for i in range(start, n):
                curr.append(candidates[i])
                backtracking(i)
                curr.pop()

        backtracking(0)

        return ans